src/kde/test/tstkde.cc

```diff
@@ -0,0 +1,1107 @@
+      dd_position_array[i] = position_array[i + rtt_c4::node() * 3];
+      dd_bandwidth_array[i] = bandwidth_array[i + rtt_c4::node() * 3];
+    }
+
+    const bool dd = true;
+    std::vector<double> smooth_result =
+        test_kde.reconstruction<1>(dd_data, dd_position_array, dd_bandwidth_array, dd);
+
+    for (int i = 0; i < local_size; i++) {
+      if (!rtt_dsxx::soft_equiv(smooth_result[i], 0.0, 1e-2))
+        ITFAILS;
+    }
+
+    double smooth_conservation = std::accumulate(smooth_result.begin(), smooth_result.end(), 0.0);
+    rtt_c4::global_sum(smooth_conservation);
+
+    // Energy conservation
+    if (!rtt_dsxx::soft_equiv(std::accumulate(data.begin(), data.end(), 0.0), smooth_conservation))
+      ITFAILS;
+  }
+
+  // what if half of it is negative and the mean is zero for a mean reconstruction
+  {
+    std::vector<double> data{-0.2, 0.2, -0.2, 0.2, -0.2, 0.2, -0.2, 0.2, -0.2, 0.2};
+    std::vector<std::array<double, 3>> position_array(10, std::array<double, 3>{0.0, 0.0, 0.0});
+    for (int i = 0; i < 10; i++) {
+      position_array[i][0] = i < 5 ? i % 5 : i % 5 + 0.5;
+      position_array[i][1] = i < 5 ? 0.5 : -0.5;
+    }
+    std::vector<std::array<double, 3>> bandwidth_array(10, std::array<double, 3>{4.0, 0., 0.0});
+
+    // map to dd arrays with simple stride
+    std::vector<double> dd_data(local_size, 0.0);
+    std::vector<std::array<double, 3>> dd_position_array(local_size,
+                                                         std::array<double, 3>{0.0, 0.0, 0.0});
+    std::vector<std::array<double, 3>> dd_bandwidth_array(local_size,
+                                                          std::array<double, 3>{0.0, 0., 0.0});
+
+    for (int i = 0; i < local_size; i++) {
+      dd_data[i] = data[i + rtt_c4::node() * 3];
+      dd_position_array[i] = position_array[i + rtt_c4::node() * 3];
+      dd_bandwidth_array[i] = bandwidth_array[i + rtt_c4::node() * 3];
+    }
+
+    const bool dd = true;
+    std::vector<double> smooth_result =
+        test_kde.mean_reconstruction<1>(dd_data, dd_position_array, dd_bandwidth_array, dd);
+
+    for (int i = 0; i < local_size; i++) {
+      if (!rtt_dsxx::soft_equiv(smooth_result[i], 0.0, 1e-2))
+        ITFAILS;
+    }
+
+    double smooth_conservation = std::accumulate(smooth_result.begin(), smooth_result.end(), 0.0);
+    rtt_c4::global_sum(smooth_conservation);
+
+    // Energy conservation
+    if (!rtt_dsxx::soft_equiv(std::accumulate(data.begin(), data.end(), 0.0), smooth_conservation))
+      ITFAILS;
+  }
+
+  if (ut.numFails == 0) {
+    PASSMSG("KDE DD checks pass");
+  } else {
+    FAILMSG("KDE DD checks failed");
+  }
+}
+
+//------------------------------------------------------------------------------------------------//
+int main(int argc, char *argv[]) {
+  ParallelUnitTest ut(argc, argv, release);
+  try {
+    // >>> UNIT TESTS
+    test_replication(ut);
+    if (nodes() == 3)
+      test_decomposition(ut);
+  }
+  UT_EPILOG(ut);
```

Comment:
I'm not too picky about the tests.  Just ensure that the coverage is good and that there are sufficient comments that another developer can follow your train of thought.

Reply:
I'll run this through Visual Studio on Wed. and give you specific feedback.

Reply:
Thanks. Let me know how it looks.